### PR TITLE
skills: add a name validation function

### PR DESCRIPTION
**Pull request by zyga**
All skill, slot and type names are constrained to simple ASCII identifiers, optionally with internal dashes.

### CODE REVIEW

File: skills/core.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+type Slot struct {
+	Name  string
+	Snap  string
+	Type  string
+	Attrs map[string]interface{}
+	Apps  []string
+}
+
+// Regular expression describing correct identifiers.
+var validName = regexp.MustCompile("^[a-z]([a-z0-9-]+[a-z0-9])?$")
```
Comment (chipaca) [inline skills/core.go]: as i mentioned on telegram, this is probably a bug. `a` is a valid name, `ab` is not, and `abc` is a valid name again.

Comment (chipaca) [inline skills/core.go]: also, make the group non-capturing please

**Comment by chipaca** [comment]
:+1: (in the style of howtobasic)

File: skills/core_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+func (s *CoreSuite) TestValidateName(c *C) {
+	c.Assert(ValidateName("name with space"), ErrorMatches,
+		`"name with space" is not a valid skill or slot name`)
+	c.Assert(ValidateName("name-with-trailing-dash-"), ErrorMatches,
+		`"name-with-trailing-dash-" is not a valid skill or slot name`)
+	c.Assert(ValidateName("name-with-3-dashes"), IsNil)
+	c.Assert(ValidateName("name"), IsNil)
+	c.Assert(ValidateName("a"), IsNil)
+	c.Assert(ValidateName("ab"), IsNil)
+	c.Assert(ValidateName("abc"), IsNil)
```
Comment (mvo5) [inline skills/core_test.go]: We might consider testing some more legal/illegal strings here. Might be overkill but then regex are hard and the test documents what is legal and illegal. Maybe:
```
legal := []string{"a", "ab", "a-b", "a0", "a-b-c"}
invalid := []string{"", "1", "a-", "a ", " ", "-"}
```
I.e. also testing the restriction that the first char has to be [a-z]

**Comment by mvo5** [comment]
Looks good, one suggestion for test improvements, maybe overkill but 708be5e would be caught with a stricter test. Plus it is a nice documentation for people who are lazy at reading regexp ;) But certainly no blocker.

File: skills/core_test.go (hunk 2 above)
Comment (zyga) [inline skills/core_test.go]: Good suggestion. I'll do it.

**Comment by mvo5** [comment]
Nice! :+1:

File: skills/core.go [hunk 3]
```diff
@@ -36,3 +41,15 @@ type Slot struct {
...
 }
+
+// Regular expression describing correct identifiers.
+var validName = regexp.MustCompile("^[a-z](:?[a-z0-9-]*[a-z0-9])?$")
+
+// ValidateName checks if a string can be used as a skill or slot name.
+func ValidateName(name string) error {
+	valid := validName.MatchString(name)
+	if !valid {
+		return fmt.Errorf("%q is not a valid skill or slot name", name)
```
Comment (niemeyer) [inline skills/core.go]: `"invalid skill name: %q"`

It seems okay to say that even if it's a slot.

File: skills/core.go [hunk 4]
```diff
@@ -36,3 +41,15 @@ type Slot struct {
 	Attrs map[string]interface{}
 	Apps  []string
 }
+
+// Regular expression describing correct identifiers.
+var validName = regexp.MustCompile("^[a-z](:?[a-z0-9-]*[a-z0-9])?$")
```
Comment (niemeyer) [inline skills/core.go]: I think this should be: `"^[a-z](?:-?[a-z0-9])*$"`

No double dashes (needs test).

**Pull request merged**